redis/connection.py

```diff
@@ -734,6 +715,188 @@ def _host_error(self):
+    else:
+        raise TypeError("Key must be either a string or bytes")
+
+
+class CacheProxyConnection(ConnectionInterface):
+    def __init__(
+        self,
+        conn: ConnectionInterface,
+        cache: Cache,
+        conf: CacheConfiguration,
+        cache_lock: threading.Lock,
+    ):
+        self.pid = os.getpid()
+        self._conn = conn
+        self.retry = self._conn.retry
+        self.host = self._conn.host
+        self.port = self._conn.port
+        self._cache = cache
+        self._cache_lock = cache_lock
+        self._conf = conf
+        self._current_command_hash = None
+        self._current_command_keys = None
+        self._current_options = None
+        self._keys_mapping = LRUCache(maxsize=10000)
+        self.register_connect_callback(self._enable_tracking_callback)
+
+    def repr_pieces(self):
+        return self._conn.repr_pieces()
+
+    def register_connect_callback(self, callback):
+        self._conn.register_connect_callback(callback)
+
+    def deregister_connect_callback(self, callback):
+        self._conn.deregister_connect_callback(callback)
+
+    def set_parser(self, parser_class):
+        self._conn.set_parser(parser_class)
+
+    def connect(self):
+        self._conn.connect()
+
+    def on_connect(self):
+        self._conn.on_connect()
+
+    def disconnect(self, *args):
+        with self._cache_lock:
+            self._cache.clear()
+        self._conn.disconnect(*args)
+
+    def check_health(self):
+        self._conn.check_health()
+
+    def send_packed_command(self, command, check_health=True):
+        # TODO: Investigate if it's possible to unpack command
+        #  or extract keys from packed command
+        self._conn.send_packed_command(command)
+
+    def send_command(self, *args, **kwargs):
+        self._process_pending_invalidations()
+
+        # If command is write command or not allowed to cache
+        # transfer control to the actual connection.
+        if not self._conf.is_allowed_to_cache(args[0]):
+            self._current_command_hash = None
+            self._current_command_keys = None
+            self._conn.send_command(*args, **kwargs)
+            return
+
+        # Create hash representation of current executed command.
+        self._current_command_hash = hashkey(*args)
+
+        # Extract keys from current command.
+        if kwargs.get("keys"):
+            self._current_command_keys = kwargs["keys"]
+
+            if not isinstance(self._current_command_keys, list):
+                raise TypeError("Cache keys must be a list.")
+
+        with self._cache_lock:
+            # If current command reply already cached
+            # prevent sending data over socket.
+            if self._cache.get(self._current_command_hash):
+                return
+
+            # Set temporary entry as a status to prevent
+            # race condition from another connection.
+            self._cache[self._current_command_hash] = "caching-in-progress"
```

Comment:
`"caching-in-progress"` Should be defined as constant.

Reply:
Fixed